### PR TITLE
[SPARK-27392][SQL] TestHive test tables should be placed in shared test state, not per session

### PR DESCRIPTION
**Pull request by ericl**
## What changes were proposed in this pull request?

Otherwise, tests that use tables from multiple sessions will run into issues if they access the same table. The correct location is in shared state.

A couple other minor test improvements.

cc @gatorsmile @srinathshankar 

## How was this patch tested?

Existing unit tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/QueryExecutionMetering.scala [hunk 1]
```diff
@@ -64,10 +64,13 @@ case class QueryExecutionMetering() {
   /** Dump statistics about time spent running specific rules. */
   def dumpTimeSpent(): String = {
     val map = timeMap.asMap().asScala
-    val maxLengthRuleNames = if (map.isEmpty) {
-      0
-    } else {
-      map.keys.map(_.toString.length).max
+    val maxLengthRuleNames = {
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/QueryExecutionMetering.scala]: Did this fix anything? Looks like equivalent logic and would be ever so slightly slower

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by xuanyuanking** [comment]
retest this please.

**Comment by xuanyuanking** [comment]
Hi @ericl, thanks for the fix, the issues of accessing the same table means getting a wrong answer or an exception? It seems that I met the same problem before.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/QueryExecutionMetering.scala (hunk 1 above)
Comment (ericl) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/QueryExecutionMetering.scala]: Oops, looks like it was already fixed in master.

**Comment by ericl** [comment]
@xuanyuanking this would only cause an exception trying to load the same test table twice.

**Comment by xuanyuanking** [comment]
retest this please.

**Review by xuanyuanking — APPROVED** [review]
Copy that, thanks Eric. The UT failure is not related.

File: mllib/src/test/java/org/apache/spark/mllib/regression/JavaRidgeRegressionSuite.java [hunk 2]
```diff
@@ -55,7 +56,8 @@ public void runRidgeRegressionUsingConstructor() {
     int numFeatures = 20;
     List<LabeledPoint> data = generateRidgeData(2 * numExamples, numFeatures, 10.0);
 
-    JavaRDD<LabeledPoint> testRDD = jsc.parallelize(data.subList(0, numExamples));
+    JavaRDD<LabeledPoint> testRDD = jsc.parallelize(
+      new ArrayList<LabeledPoint>(data.subList(0, numExamples)));
```
Comment (srowen) [inline mllib/src/test/java/org/apache/spark/mllib/regression/JavaRidgeRegressionSuite.java]: Why this change? I don't see how it relates, or why a view of the list would be an issue here.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 3]
```diff
@@ -480,14 +484,12 @@ private[hive] class TestHiveSparkSession(
...
   }
 
-  private val loadedTables = new collection.mutable.HashSet[String]
-
-  def getLoadedTables: collection.mutable.HashSet[String] = loadedTables
+  def getLoadedTables: collection.mutable.HashSet[String] = sharedState.loadedTables
 
   def loadTestTable(name: String) {
-    if (!(loadedTables contains name)) {
+    if (!(sharedState.loadedTables contains name)) {
```
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: PS I wouldn't mind avoiding postfix notation here. I'm trying to remove it from the code base.

File: mllib/src/test/java/org/apache/spark/mllib/regression/JavaRidgeRegressionSuite.java (hunk 2 above)
Comment (srowen) [reply mllib/src/test/java/org/apache/spark/mllib/regression/JavaRidgeRegressionSuite.java]: CC @ericl

Comment (ericl) [reply mllib/src/test/java/org/apache/spark/mllib/regression/JavaRidgeRegressionSuite.java]: The issue here is that list views are not serializable. This fixes the issue, and does not affect the behaviour of the test.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala (hunk 3 above)
Comment (ericl) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: Fixed

File: mllib/src/test/java/org/apache/spark/mllib/regression/JavaRidgeRegressionSuite.java (hunk 2 above)
Comment (srowen) [reply mllib/src/test/java/org/apache/spark/mllib/regression/JavaRidgeRegressionSuite.java]: Ah OK. Kind of unrelated, but OK to fix if it came up while working on the issue.

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master
